// sonarqube-scanner/src/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheStatus = exports.ScannerProperty = void 0;
var ScannerProperty;
(function (ScannerProperty) {
    ScannerProperty["SonarVerbose"] = "sonar.verbose";
    ScannerProperty["SonarLogLevel"] = "sonar.log.level";
    ScannerProperty["SonarToken"] = "sonar.token";
    ScannerProperty["SonarExclusions"] = "sonar.exclusions";
    ScannerProperty["SonarHostUrl"] = "sonar.host.url";
    ScannerProperty["SonarUserHome"] = "sonar.userHome";
    ScannerProperty["SonarWorkingDirectory"] = "sonar.working.directory";
    ScannerProperty["SonarScannerApiBaseUrl"] = "sonar.scanner.apiBaseUrl";
    ScannerProperty["SonarScannerOs"] = "sonar.scanner.os";
    ScannerProperty["SonarScannerArch"] = "sonar.scanner.arch";
    ScannerProperty["SonarOrganization"] = "sonar.organization";
    ScannerProperty["SonarProjectBaseDir"] = "sonar.projectBaseDir";
    ScannerProperty["SonarScannerSonarCloudUrl"] = "sonar.scanner.sonarcloudUrl";
    ScannerProperty["SonarScannerJavaExePath"] = "sonar.scanner.javaExePath";
    ScannerProperty["SonarScannerJavaOptions"] = "sonar.scanner.javaOpts";
    ScannerProperty["SonarScannerWasJreCacheHit"] = "sonar.scanner.wasJreCacheHit";
    ScannerProperty["SonarScannerWasEngineCacheHit"] = "sonar.scanner.wasEngineCacheHit";
    ScannerProperty["SonarScannerProxyHost"] = "sonar.scanner.proxyHost";
    ScannerProperty["SonarScannerProxyPort"] = "sonar.scanner.proxyPort";
    ScannerProperty["SonarScannerProxyUser"] = "sonar.scanner.proxyUser";
    ScannerProperty["SonarScannerProxyPassword"] = "sonar.scanner.proxyPassword";
    ScannerProperty["SonarScannerResponseTimeout"] = "sonar.scanner.responseTimeout";
    ScannerProperty["SonarScannerSkipJreProvisioning"] = "sonar.scanner.skipJreProvisioning";
    ScannerProperty["SonarScannerInternalDumpToFile"] = "sonar.scanner.internal.dumpToFile";
    ScannerProperty["SonarScannerTruststorePath"] = "sonar.scanner.truststorePath";
    ScannerProperty["SonarScannerKeystorePath"] = "sonar.scanner.keystorePath";
    ScannerProperty["SonarScannerKeystorePassword"] = "sonar.scanner.keystorePassword";
    ScannerProperty["SonarScannerTruststorePassword"] = "sonar.scanner.truststorePassword";
    ScannerProperty["SonarScannerInternalIsSonarCloud"] = "sonar.scanner.internal.isSonarCloud";
    ScannerProperty["SonarScannerInternalSqVersion"] = "sonar.scanner.internal.sqVersion";
    ScannerProperty["SonarScannerCliVersion"] = "sonar.scanner.version";
    ScannerProperty["SonarScannerCliMirror"] = "sonar.scanner.mirror";
    // Deprecated properties:
    ScannerProperty["SonarWsTimeout"] = "sonar.ws.timeout";
    ScannerProperty["HttpProxyHost"] = "http.proxyHost";
    ScannerProperty["HttpProxyPort"] = "http.proxyPort";
    ScannerProperty["HttpProxyUser"] = "http.proxyUser";
    ScannerProperty["HttpProxyPassword"] = "http.proxyPassword";
    ScannerProperty["SonarLogin"] = "sonar.login";
})(ScannerProperty || (exports.ScannerProperty = ScannerProperty = {}));
var CacheStatus;
(function (CacheStatus) {
    CacheStatus["Hit"] = "hit";
    CacheStatus["Miss"] = "miss";
    CacheStatus["Disabled"] = "disabled";
})(CacheStatus || (exports.CacheStatus = CacheStatus = {}));
